forui/lib/src/widgets/pagination/pagination_controller.dart

```diff
@@ -0,0 +1,192 @@
+  final int _length;
+  int _value;
+
+  /// Creates a [FPaginationController].
+  ///
+  /// # Contract:
+  /// * Throws [AssertionError] if 1 <= [page] and [page] <= length.
+  FPaginationController({
+    int? length,
+    int page = 0,
+    this.showEdges = true,
+    this.siblings = 1,
+    super.onAttach,
+    super.onDetach,
+    super.keepPage,
+    super.viewportFraction,
+  }) : assert(length == null || 0 < length, 'The total length of pages should be more than 0, but is $length.'),
+       assert(0 <= siblings, 'The siblingLength should be non-negative, but is $siblings'),
+       assert(
+         0 <= page && page < (length ?? 0),
+         'The initial page must be greater than or equal to 0 and less than or equal to length.',
+       ),
+       _value = page,
+       _length = length ?? 0,
+       super(initialPage: page) {
+    addListener(_handleScrollChanges);
+  }
+
+  void _handleScrollChanges() {
+    if (hasClients) {
+      if (page != null && page!.round() != _value) {
+        _value = page!.round();
+        notifyListeners();
+      }
+    }
+  }
+
+  /// Returns the total number of pages in the pagination.
+  ///
+  /// If the controller has clients (i.e., it is attached to a [PageView]),
+  /// the length is dynamically computed using the scroll extents and viewport dimensions.
+  ///
+  /// If not attached to any clients, it returns the predefined `_length`, which defaults to 0.
+  num get length {
+    if (hasClients) {
+      final len =
+          max(0, position.maxScrollExtent - position.minScrollExtent) /
+          max(1.0, position.viewportDimension * viewportFraction);
+      if (len.isFinite) {
+        return len.ceil() + 1;
+      } else {
+        return len;
+      }
+    } else {
+      return _length;
+    }
+  }
+
+  /// Moves to the previous page if the current page is greater than 1.
+  void previous() {
+    final value = hasClients && page != null ? page!.round() : _value;
+    if (0 < value) {
+      _value = value - 1;
+
+      if (hasClients) {
+        super.previousPage(duration: const Duration(milliseconds: 300), curve: Curves.ease);
+      }
+      notifyListeners();
+    }
+  }
+
+  /// Moves to the next page if the current page is less than the total number of pages.
+  void next() {
+    final value = hasClients && page != null ? page!.round() : _value;
+    if (value < length - 1) {
+      _value = value + 1;
+
+      if (hasClients) {
+        super.nextPage(duration: const Duration(milliseconds: 300), curve: Curves.ease);
+      }
+      notifyListeners();
+    }
+  }
+
+  /// Returns the current page index.
+  int get value => _value;
+
+  /// Sets the current page index.
+  set value(int index) {
+    if (0 <= index && index <= length) {
+      _value = index;
+
+      if (hasClients) {
+        super.jumpToPage(index);
+      }
+      notifyListeners();
+    } else {
+      throw StateError('The index must be within the allowed range.');
+    }
+  }
```

Comment:
I think we need to be consistent with the terminology we use. What differentiates a page from a value?

Reply:
The reason I differentiated the 2 is because `page` is firstly already used as a getter by `PageController` to get the current page.  `value` represents the current index of the pagination.